### PR TITLE
blockchain/v1: fix deadlock

### LINKED ISSUE #4591: Possible deadlock in StopPeerForError -> RemovePeer
**Tendermint version** (use `tendermint version` or `git rev-parse --verify HEAD` if installed from source): v0.33.2

**ABCI app** (name for built-in, URL for self-written if it's publicly available): https://github.com/MinterTeam/minter-go-node

**Environment**:
- **OS** (e.g. from /etc/os-release): Debian GNU/Linux 10 (buster)

**What happened**: Node cannot be terminated by `SIGTERM` after some time of operating. Many `chan send`-ing goroutines found in `kill -6` output. 

Examples (there are 100+ of them in logs):
```
goroutine 122508 [chan send, 2320 minutes]:
github.com/tendermint/tendermint/blockchain/v1.(*BlockchainReactor).RemovePeer(0xc03ad5a600, 0x14098c0, 0xc06b9bb5c0, 0xf6dca0, 0xc00007e0f0)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/blockchain/v1/reactor.go:211 +0xef
```

```
goroutine 3759947 [chan send, 2330 minutes]:
github.com/tendermint/tendermint/blockchain/v1.(*BlockchainReactor).RemovePeer(0xc03ad5a600, 0x14098c0, 0xc026134a80, 0xf6dca0, 0xc00007e0f0)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/blockchain/v1/reactor.go:211 +0xef
github.com/tendermint/tendermint/p2p.(*Switch).stopAndRemovePeer(0xc0002d57a0, 0x14098c0, 0xc026134a80, 0xf6dca0, 0xc00007e0f0)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/p2p/switch.go:356 +0x11c
github.com/tendermint/tendermint/p2p.(*Switch).StopPeerForError(0xc0002d57a0, 0x14098c0, 0xc026134a80, 0xf6dca0, 0xc00007e0f0)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/p2p/switch.go:325 +0x127
github.com/tendermint/tendermint/p2p.createMConnection.func2(0xf6dca0, 0xc00007e0f0)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/p2p/peer.go:390 +0x4e
github.com/tendermint/tendermint/p2p/conn.(*MConnection).stopForError(0xc05aad5600, 0xf6dca0, 0xc00007e0f0)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/p2p/conn/connection.go:341 +0x76
github.com/tendermint/tendermint/p2p/conn.(*MConnection).recvRoutine(0xc05aad5600)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/p2p/conn/connection.go:596 +0xd2d
created by github.com/tendermint/tendermint/p2p/conn.(*MConnection).OnStart
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/p2p/conn/connection.go:230 +0x1d9
```

TrapSignal routine is also asleep:
```
goroutine 743 [chan send]:
github.com/tendermint/tendermint/blockchain/v1.(*BlockchainReactor).RemovePeer(0xc03ad5a600, 0x14098c0, 0xc05aac0f00, 0x0, 0x0)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/blockchain/v1/reactor.go:211 +0xef
github.com/tendermint/tendermint/p2p.(*Switch).stopAndRemovePeer(0xc0002d57a0, 0x14098c0, 0xc05aac0f00, 0x0, 0x0)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/p2p/switch.go:356 +0x11c
github.com/tendermint/tendermint/p2p.(*Switch).OnStop(0xc0002d57a0)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/p2p/switch.go:243 +0x9f
github.com/tendermint/tendermint/libs/service.(*BaseService).Stop(0xc0002d57a0, 0x0, 0x0)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/libs/service/service.go:167 +0x2f4
github.com/tendermint/tendermint/node.(*Node).OnStop(0xc03aea6000)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/node/node.go:816 +0xb9
github.com/tendermint/tendermint/libs/service.(*BaseService).Stop(0xc03aea6000, 0x4000, 0xc089b763c0)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/libs/service/service.go:167 +0x2f4
github.com/MinterTeam/minter-go-node/cmd/minter/cmd.runNode.func4()
	/go/src/github.com/MinterTeam/minter-go-node/cmd/minter/cmd/node.go:109 +0x4c
github.com/tendermint/tendermint/libs/os.TrapSignal.func1(0xc03a5e7080, 0x7f9aa292df60, 0xc039cb94e0, 0xc039cb9500)
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/libs/os/os.go:24 +0x109
created by github.com/tendermint/tendermint/libs/os.TrapSignal
	/go/pkg/mod/github.com/tendermint/tendermint@v0.33.2/libs/os/os.go:20 +0xe6
``` 

**What you expected to happen**: Node should terminate correctly.

**Have you tried the latest version**: yes

### PR DESCRIPTION
**Pull request by melekes**
I introduced a new variable - syncEnded, which is now used to prevent
sending new events to channels (which would block otherwise) if reactor
is finished syncing

Closes #4591

### CODE REVIEW

File: blockchain/v1/reactor.go [hunk 1]
```diff
@@ -67,6 +68,9 @@ type BlockchainReactor struct {
 	eventsFromFSMCh chan bcFsmMessage
 
 	swReporter *behaviour.SwitchReporter
+
+	// Atomic integer (0 - sync in progress, 1 - finished syncing)
+	syncEnded int32
```
Comment (alexanderbez) [inline blockchain/v1/reactor.go:73]: why not just store a reference here?

Comment (melekes) [reply blockchain/v1/reactor.go:73]: what are the benefits?

Comment (alexanderbez) [reply blockchain/v1/reactor.go:73]: You don't need to do `&(bcR.syncEnded)`

Comment (melekes) [reply blockchain/v1/reactor.go:73]: But then I will need to set initial value of 0 in constructor. I prefer my version for that reason.

Comment (alexanderbez) [reply blockchain/v1/reactor.go:73]: but zero is the default value. What would you explicitly need to set it to?

Comment (alexanderbez) [reply blockchain/v1/reactor.go:73]: This isn't blocking, but given that you continuously need to dereference, I think it makes it harder to read. Up to you :)

Comment (melekes) [reply blockchain/v1/reactor.go:73]: > but zero is the default value.

not if it's a pointer! nil is zero value for pointer

**Review by erikgrinaker — APPROVED** [review]
I feel like we should probably use the existing termination mechanism rather than introducing additional ones, but since we're removing this anyway I suppose this is fine.

File: blockchain/v1/reactor.go (hunk 1 above)
Comment (alexanderbez) [reply blockchain/v1/reactor.go:73]: No, I meant it's zero if you make it an `int32` (per my suggestion). So my question is, is zero a (sync in progress) the correct default value?

Comment (melekes) [reply blockchain/v1/reactor.go:73]: yes, 0 is correct default value

Comment (alexanderbez) [reply blockchain/v1/reactor.go:73]: > But then I will need to set initial value of 0 in constructor.

Then you don't need to do this.

**Pull request merged**